### PR TITLE
introduce "systemd-analyze capability" command for dumping locally defined caps

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/analyze/analyze.c [hunk 1]
```diff
@@ -1592,6 +1594,47 @@ static int dump_exit_status(int argc, char *argv[], void *userdata) {
...
+                                           TABLE_STRING, capability_to_name(c) ?: "cap_???",
+                                           TABLE_UINT, c);
+                        if (r < 0)
+                                return table_log_add_error(r);
+                }
+        else {
+                for (int i = 1; i < argc; i++) {
+                        int c;
+
+                        c = capability_from_name(argv[i]);
```
Comment (yuwata) [inline src/analyze/analyze.c:1623]: If a large numeric value is specified, and both kernel and systemd do not know the corresponding capability, then it still shows `cap_???`. I think we need to check the value is less or equal to `MAX(CAP_LAST_CAP, cap_last_cap())`.

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -544,22 +544,26 @@
...
+        listed capabilities will be included, the effect of the assignment inverted. Note that this option
+        also affects the respective capabilities in the effective, permitted and inheritable capability
+        sets. If this option is not used, the capability bounding set is not modified on process execution,
+        hence no limits on the capabilities of the process are enforced. This option may appear more than
+        once, in which case the bounding sets are merged by <constant>OR</constant>, or by
+        <constant>AND</constant> if the lines are prefixed with <literal>~</literal> (see below). If the
+        empty string is assigned to this option, the bounding set is reset to the empty capability set, and
+        all prior settings have no effect.  If set to <literal>~</literal> (without any further argument),
+        the bounding set is reset to the full set of available capabilities, also undoing any previous
+        settings. This does not affect commands prefixed with <literal>+</literal>. Use
```
Comment (keszybz) [inline man/systemd.exec.xml]: This would be clearer with a a separate paragraph I think.

File: src/analyze/analyze.c (hunk 1 above)
Comment (keszybz) [reply src/analyze/analyze.c:1623]: I'm not sure. It seems valid to ask for a capability that is defined in the upstream kernel on an older systemd running on an older kernel. And the code already refuses caps above 63 which we cannot encode. So I think this is OK as is.

**Review by keszybz — APPROVED** [review]
LGTM.

File: src/analyze/analyze.c (hunk 1 above)
Comment (yuwata) [reply src/analyze/analyze.c:1623]: Currently, 40 (cap_checkpoint_restore) is the largest known capability. But, if, e.g. `systemd-analyze capability 60`, then it shows `cap_??? 60`. If this is acceptable, then man page needs to be updated. Currently, it is documented that `cap_???` is shown only when the capability is known by kernel but not by systemd.

File: man/systemd-analyze.xml [hunk 3]
```diff
@@ -345,6 +351,30 @@ DATAERR 65     BSD
       </example>
     </refsect2>
 
+    <refsect2>
+      <title><command>systemd-analyze capability <optional><replaceable>CAPABILITY</replaceable>...</optional></command></title>
+
+      <para>This command prints a list of Linux capabilities along with their numeric IDs. See <citerefentry
+      project='man-pages'><refentrytitle>capabilities</refentrytitle><manvolnum>7</manvolnum></citerefentry>
+      for details. If no argument is specified the full list of capabilities known to the service manager and
+      the kernel is shown. Capabilities defined by the kernel but not known to the service manager are shown
```
Comment (keszybz) [inline man/systemd-analyze.xml:360]: @yuwata has a point: please strike "defined by the kernel but"

**Review by keszybz — CHANGES_REQUESTED** [review]

File: man/systemd-analyze.xml (hunk 3 above)
Comment (keszybz) [reply man/systemd-analyze.xml:360]: (or do as @yuwata suggests, but please resolve the discrepancy one way or another...)

File: src/analyze/analyze.c (hunk 1 above)
Comment (poettering) [reply src/analyze/analyze.c:1623]: will fix

**Comment by poettering** [comment]
made the suggested changes, ptal

**Comment by keszybz** [comment]
```
./src/analyze/analyze.c: In function ‘dump_capabilities’:
../src/analyze/analyze.c:1624:40: error: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Werror=sign-compare]
 1624 |                         if (c < 0 || c > last_cap)
      |                                        ^
cc1: all warnings being treated as errors
```

**Pull request merged**